### PR TITLE
[WIP] Add checking with valgrind

### LINKED ISSUE #64: Add fsanitize testing
Found some issues in #63. Plus we would want to satisfy an optional requirement of xSDK (cf #58).

### PR DESCRIPTION
**Pull request by masterleinad**
Fixes #64.

### CODE REVIEW

**Comment by aprokop** [comment]
So `memcheck` runs but produces:
```
-- Processing memory checking output:
2/14 MemCheck: #2: ArborX_DetailsUtils .......................   Defects: 8
4/14 MemCheck: #4: ArborX_LinearBVH ..........................   Defects: 8
5/14 MemCheck: #5: ArborX_DetailsTreeConstruction ............   Defects: 8
7/14 MemCheck: #7: ArborX_DetailsTreeTraversal ...............   Defects: 8
8/14 MemCheck: #8: ArborX_DetailsBatchedQueries ..............   Defects: 8
9/14 MemCheck: #9: ArborX_DistributedSearchTree ..............   Defects: 174
10/14 MemCheck: #10: ArborX_DetailsDistributedSearchTreeImpl ...   Defects: 8
11/14 MemCheck: #11: ArborX_BoostAdapters ......................   Defects: 8
12/14 MemCheck: #12: ArborX_BoundingVolumeHierarchy_Example ....   Defects: 11
13/14 MemCheck: #13: ArborX_DistributedTree_Example ............   Defects: 174
14/14 MemCheck: #14: ArborX_TreeViz_Example ....................   Defects: 8
MemCheck log files can be found here: ( * corresponds to test number)
/var/jenkins/workspace/ArborX-experimental_PR-67/build/Testing/Temporary/MemoryChecker.*.log
Memory checking results:
Memory Leak - 293
Potential Memory Leak - 121
Invalid syscall param - 9
```

**Comment by masterleinad** [comment]
Locally, I see the usual `OpenMPI` problems when running `valgrind` on the `MPI` tests (but only for them). I am not sure what the problem with the other tests we see here is.

**Comment by masterleinad** [comment]
We get a lot of
```
==1037== 1,360 bytes in 10 blocks are possibly lost in loss record 1,230 of 1,489
==1037==    at 0x4C31B25: calloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==1037==    by 0xA7A7D02: ??? (in /usr/lib/x86_64-linux-gnu/libcuda.so.418.67)
==1037==    by 0xA7A8330: ??? (in /usr/lib/x86_64-linux-gnu/libcuda.so.418.67)
==1037==    by 0xA7A88E4: ??? (in /usr/lib/x86_64-linux-gnu/libcuda.so.418.67)
==1037==    by 0xA80C146: ??? (in /usr/lib/x86_64-linux-gnu/libcuda.so.418.67)
==1037==    by 0xA80C3E7: ??? (in /usr/lib/x86_64-linux-gnu/libcuda.so.418.67)
==1037==    by 0xA5F57E9: ??? (in /usr/lib/x86_64-linux-gnu/libcuda.so.418.67)
==1037==    by 0xA6D99CE: ??? (in /usr/lib/x86_64-linux-gnu/libcuda.so.418.67)
==1037==    by 0xA700632: ??? (in /usr/lib/x86_64-linux-gnu/libcuda.so.418.67)
==1037==    by 0xA63DA97: ??? (in /usr/lib/x86_64-linux-gnu/libcuda.so.418.67)
==1037==    by 0xA63E183: ??? (in /usr/lib/x86_64-linux-gnu/libcuda.so.418.67)
==1037==    by 0x53361E6: ??? (in /usr/local/cuda-9.2/targets/x86_64-linux/lib/libcudart.so.9.2.148)
```

**Comment by dalg24** [comment]
We can add a suppression file for these

File: test/CMakeLists.txt [hunk 1]
```diff
@@ -1,3 +1,12 @@
+IF(ArborX_TEST_MEMORY)
+  FIND_PROGRAM(MEMORYCHECK_COMMAND NAMES valgrind)
+  IF(MEMORYCOMMAND MATCHES "-NOTFOUND")
+    MESSAGE(FATAL_ERROR
+	    "Memory checking in tests enabled but could not find the valgrind "
+	    "executable. Please install valgrind.")
+  ENDIF()
+  SET(CTEST_MEMORYCHECK_COMMAND "${MEMORYCHECK_COMMAND}")
```
Comment (dalg24) [inline test/CMakeLists.txt]: Are you sure this variable is needed at all if we do `ctest -T MemCheck`?

Comment (masterleinad) [reply test/CMakeLists.txt]: I thought so, but it seems to run without it as well now.

**Comment by masterleinad** [comment]
Documenting:
```
export LSAN_OPTIONS=suppressions=/tmp/ArborX/memory_blacklist.txt
```
where `memory_blacklist.txt` contains
```
leak:libmpi.so (effect?)
leak:libopen-pal.so
leak:exe (effect?)
```

**Comment by aprokop** [comment]
Do we want to merge this?

**Comment by aprokop** [comment]
Closing as no there has been no activity or interest in the past 8 months. Reopen as needed.

**Pull request closed without merging**